hibernate-core/src/main/java/org/hibernate/query/sqm/tree/expression/AbstractSqmExpression.java

```diff
@@ -156,4 +156,8 @@ public JpaSelection<T> alias(String name) {
 	public JavaType<T> getJavaTypeDescriptor() {
 		return getNodeType() == null ? null : getNodeType().getExpressibleJavaType();
 	}
+
+	public void forceInferableType(SqmExpressible<?> type){
```

Comment:
I think we should move the logic from `org.hibernate.query.internal.QueryParameterBindingImpl#setType` to `AbstractSqmParameter#applyInferableType` and call that method instead. Changing the type of the parameter binding is a no-go AFAIU from our last discussion, so we shouldn't do it and deprecate the `setType` method.